elastic/kibana
# [SO migrations] exit early if cluster routing allocation is disabled

## Summary
Resolves https://github.com/elastic/kibana/issues/124139.

Upgrade migrations will fail when replica allocation is disabled (`cluster.routing.allocation.enable: none`) because the REINDEX_SOURCE_TO_TEMP_INDEX_BULK step performs a bulk index operation with `wait_for_active_shards=all`. 
This PR adds a check to see if replication is enabled during the migrations initialization step and will fail the migration early if replica allocation is disabled.

### Logging
With logging configured as:
```yaml
//kibana.yml
logging:
  appenders:
    console:
      type: console
      layout:
        type: pattern
        highlight: true
        pattern: "[%date][%level][%logger]---%message"
    json:
      type: console
      layout:
        type: json
  loggers:
    - name: root
      level: info
      appenders: [console, json]
```
If routing allocation is set to "primaries", "new_primaries" or "none", Kibana will fail to start and will log the 
<details>
  <summary>following message:</summary>

```json
{"ecs":{"version":"8.0.0"},"@timestamp":"2022-03-04T15:46:25.259-07:00","message":"Unable to complete saved object migrations for the [.kibana] index: The elasticsearch cluster has cluster routing allocation incorrectly set for migrations to continue. To proceed, please remove the cluster routing allocation settings with PUT /_cluster/settings {\"transient\": {\"cluster.routing.allocation.enable\": null}, \"persistent\": {\"cluster.routing.allocation.enable\": null}}","error":{"message":"Unable to complete saved object migrations for the [.kibana] index: The elasticsearch cluster has cluster routing allocation incorrectly set for migrations to continue. To proceed, please remove the cluster routing allocation settings with PUT /_cluster/settings {\"transient\": {\"cluster.routing.allocation.enable\": null}, \"persistent\": {\"cluster.routing.allocation.enable\": null}}","type":"Error","stack_trace":"Error: Unable to complete saved object migrations for the [.kibana] index: The elasticsearch cluster has cluster routing allocation incorrectly set for migrations to continue. To proceed, please remove the cluster routing allocation settings with PUT /_cluster/settings {\"transient\": {\"cluster.routing.allocation.enable\": null}, \"persistent\": {\"cluster.routing.allocation.enable\": null}}\n    at migrationStateActionMachine (....stack trace...)
```
</details>

With a pattern layout, the logs are 
<details>
	<summary>more concise:</summary>

```log
[2022-03-04T15:46:25.259-07:00][FATAL][root]---Error: Unable to complete saved object migrations for the [.kibana] index: The elasticsearch cluster has cluster routing allocation incorrectly set for migrations to continue. To proceed, please remove the cluster routing allocation settings with PUT /_cluster/settings {"transient": {"cluster.routing.allocation.enable": null}, "persistent": {"cluster.routing.allocation.enable": null}}
    at migrationStateActionMachine (..stack trace...)
```
</details>

Elasticsearch logs should contain corresponding logs when routing allocation is changed as a transient or persistent setting:
```
   │ info [o.e.c.s.ClusterSettings] [ChristianesMBP2] updating [cluster.routing.allocation.enable] from [all] to [none]
```

### How to test this
- Configure logging using the configuration shown in the description of this PR. You can either use `json` or `pattern` layout.
- Start up elasticsearch and kibana
- With Kibana running, from dev tools, update the cluster settings to set `routing.allocation.enable` to one of 'none', 'primaries' or 'new_primaries' as either a transient or persistent setting. For example:
```
PUT /_cluster/settings
{
  "persistent": {
        "cluster.routing.allocation.enable" : "primaries"
  }
}
```
- Without stopping Elasticsearch, restart Kibana
- Observe that Kibana fails to start during the saved objects migrations phase with a message of 

```
 FATAL  Error: Unable to complete saved object migrations for the [.kibana] index: The elasticsearch cluster has cluster routing allocation incorrectly set for migrations to continue. To proceed, please remove the cluster routing allocation settings with PUT /_cluster/settings {"transient": {"cluster.routing.allocation.enable": null}, "persistent": {"cluster.routing.allocation.enable": null}}
```

### Checklist

- [X] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials (saved objects README updated)

### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Kibana fails to start early in the migrations phase. Before this PR, migrations would timeout if routing allocation is disabled, without more information on what the cause was. | Medium | Low | Integration tests will verify that migrations fail early with a helpful message stating why Kibana fails to start |

### For maintainers
